File: homework03-go/life_gui.go
package main

import (
	"./life"
	"github.com/akamensky/argparse"
	"github.com/hajimehoshi/ebiten"
	"github.com/hajimehoshi/ebiten/inpututil"
	"image/color"
	"log"
	"math"
	"os"
)

type GUI struct {
	life     *life.GameOfLife
	width    int
	height   int
	cellSize int
	paused   bool
}

func (gui *GUI) init(life *life.GameOfLife, cellSize int) {
	gui.life = life
	gui.width = gui.life.Cols * cellSize
	gui.height = gui.life.Rows * cellSize
	gui.cellSize = cellSize
	gui.paused = false
}

func (gui *GUI) drawLines(screen *ebiten.Image) {
	verticalLine, _ := ebiten.NewImage(1, gui.height, ebiten.FilterDefault)
	_ = verticalLine.Fill(color.Black)

	horizontalLine, _ := ebiten.NewImage(gui.width, 1, ebiten.FilterDefault)
	_ = horizontalLine.Fill(color.Black)

	for x := 0; x < gui.width; x += gui.cellSize {
		lineOpts := &ebiten.DrawImageOptions{}
		lineOpts.GeoM.Translate(float64(x), 0)
		_ = screen.DrawImage(verticalLine, lineOpts)
	}
	for y := 0; y < gui.height; y += gui.cellSize {
		lineOpts := &ebiten.DrawImageOptions{}
		lineOpts.GeoM.Translate(0, float64(y))
		_ = screen.DrawImage(horizontalLine, lineOpts)
	}
}

func (gui *GUI) drawGrid(screen *ebiten.Image) {
	cell, _ := ebiten.NewImage(gui.cellSize-1, gui.cellSize-1, ebiten.FilterDefault)
	_ = cell.Fill(color.RGBA{G: 255, A: 255})

	for y, row := range gui.life.CurrentGeneration {
		for x, cellValue := range row {
			if cellValue {
				lineOpts := &ebiten.DrawImageOptions{}
				lineOpts.GeoM.Translate(float64(x*gui.cellSize+1), float64(y*gui.cellSize+1))
				_ = screen.DrawImage(cell, lineOpts)
			}
		}
	}
}

func (gui *GUI) getCell(x, y int) (int, int) {
	return x / gui.cellSize, y / gui.cellSize
}

func (gui *GUI) run(screen *ebiten.Image) error {
	if inpututil.IsKeyJustPressed(ebiten.KeySpace) {
		gui.paused = !gui.paused
	}
	if inpututil.IsMouseButtonJustPressed(ebiten.MouseButtonLeft) && gui.paused {
		xPos, yPos := ebiten.CursorPosition()
		x, y := gui.getCell(xPos, yPos)
		gui.life.CurrentGeneration[y][x] = !gui.life.CurrentGeneration[y][x]
	}

	_ = screen.Fill(color.White)
	gui.drawLines(screen)
	gui.drawGrid(screen)

	if !gui.paused {
		gui.life.Step()
	}
	return nil
}

func main() {
	parser := argparse.NewParser("Game of Life", "Game of Life, developed by Georgy Komarov, K3141")
	cols := parser.Int("c", "columns", &argparse.Options{Default: 50})
	rows := parser.Int("r", "rows", &argparse.Options{Default: 30})
	cellSize := parser.Int("s", "cell-size", &argparse.Options{Default: 10})
	maxGen := parser.Float("m", "max-generations", &argparse.Options{Default: math.Inf(0)})
	_ = parser.Parse(os.Args)

	game := &life.GameOfLife{}
	game.Init(*rows, *cols, *maxGen)

	gui := &GUI{}
	gui.init(game, *cellSize)

	ebiten.SetRunnableInBackground(true)
	if err := ebiten.Run(gui.run, gui.width, gui.height, 1, "Game of Life - Komarov Georgy, K3141"); err != nil {
		log.Fatal(err)
	}
}

File: homework03/life-gui.py
import argparse

import pygame
from pygame.locals import *

from life import GameOfLife
from ui import UI


class GUI(UI):
    def __init__(self, life: GameOfLife, cell_size: int = 10, speed: int = 10) -> None:
        super().__init__(life)
        self.width = self.life.cols * cell_size
        self.height = self.life.rows * cell_size
        self.screen = pygame.display.set_mode((self.width, self.height))

        self.cell_size = cell_size
        self.speed = speed

    def draw_lines(self) -> None:
        """ Отрисовать сетку """
        for x in range(0, self.width, self.cell_size):
            pygame.draw.line(self.screen, pygame.Color('black'), (x, 0), (x, self.height))
        for y in range(0, self.height, self.cell_size):
            pygame.draw.line(self.screen, pygame.Color('black'), (0, y), (self.width, y))

    def draw_grid(self) -> None:
        """
        Отрисовка списка клеток с закрашиванием их в соответствующе цвета.
        """
        for y in range(len(self.life.curr_generation)):
            for x in range(len(self.life.curr_generation[0])):
                if self.life.curr_generation[y][x]:
                    pygame.draw.rect(self.screen,
                                     pygame.Color('green'),
                                     (x * self.cell_size + 1,
                                      y * self.cell_size + 1,
                                      self.cell_size - 1,
                                      self.cell_size - 1)
                                     )

    def get_cell(self, x: int, y: int):
        return x // self.cell_size, y // self.cell_size

    def run(self) -> None:
        # Copy from previous assignment
        pygame.init()
        clock = pygame.time.Clock()
        pygame.display.set_caption('Game of Life')
        self.screen.fill(pygame.Color('white'))

        paused = False
        running = True

        while running:
            for event in pygame.event.get():
                if event.type == QUIT:
                    running = False
                if event.type == KEYDOWN:
                    if event.key == K_SPACE:
                        paused = not paused
                if event.type == MOUSEBUTTONDOWN:
                    if paused:
                        x, y = self.get_cell(*pygame.mouse.get_pos())
                        self.life.curr_generation[y][x] = not self.life.curr_generation[y][x]

            self.screen.fill(pygame.Color('white'))
            self.draw_lines()
            self.draw_grid()

            # Отрисовка списка клеток
            # Выполнение одного шага игры (обновление состояния ячеек)
            if paused:
                pass
            elif not self.life.is_max_generations_exceed:
                self.life.step()

            pygame.display.flip()
            clock.tick(self.speed)
        pygame.quit()


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--width', type=int, default=640)
    parser.add_argument('--height', type=int, default=480)
    parser.add_argument('--cell-size', type=int, default=10)
    parser.add_argument('--max-generations', default=float('inf'), type=int)
    args = parser.parse_args()

    rows, cols = args.height // args.cell_size, args.width // args.cell_size
    game = GameOfLife((rows, cols), max_generations=args.max_generations)
    gui = GUI(game, args.cell_size)
    gui.run()
